### ISSUE TITLE
`winrepo.update_git_repos` runner doesn't fully restore remote state 

### ISSUE THREAD

**Issue by eliasp** [saltstack/salt#20824]
**OS:** Ubuntu 14.04.1 amd64
**Salt Master:**
```
               Salt: 2014.7.1
             Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
             Jinja2: 2.7.2
           M2Crypto: 0.21.1
     msgpack-python: 0.3.0
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.0.1
               RAET: Not Installed
                ZMQ: 4.0.4
               Mako: 0.9.1
```
**GitPython:** 0.3.2 RC1

Running the `winrepo.update_git_repos` runner should always ensure the local clone/checkout of the repos represents 100% the state of the remote.

**Relevant Master Config:**
```YAML
win_repo: /media/volumes/salt/master/fileserver/win/repo
win_gitrepos:
    - git@scm.institution.tld:salt-winrepo.git
```

The following steps show, that's not the case:
1. `salt-run winrepo.update_git_repos`
2. `find /media/volumes/salt/master/fileserver/win/repo -name "*.sls" -delete`
3. `salt-run winrepo.update_git_repos`
4. `find /media/volumes/salt/master/fileserver/win/repo -name "*.sls" | wc -l` → result: `0`

Only deleting the clone completely + `winrepo.update_git_repos` or manually doing a `git reset --hard origin/master` in the clone restores it to it's proper state.

Every run of `salt-run winrepo.update_git_repos` should ensure the clone is a 1:1 representation of the remote.

Besides that, I wonder whether it wouldn't make more sense to cache the `win_gitrepo` the same way all GitFS repos are cached instead of simply cloning it to the `win_repo` dir. This would possibly avoid quite a lot of code duplication and make for a more streamlined/unified user experience, when this could be simply hooked into `fileserver.update`.

**Comment by rallytime** [comment]
Thanks for this great report @eliasp! 

ping @terminalmage @shanedlee @UtahDave

**Comment by UtahDave** [comment]
The winrepo.update_git_repos would likely benefit tremendously by working how gitfs does. update_git_repos is very old code.

**Comment by UtahDave** [comment]
The underlying problem is the `git.latest` state that's being used by the runner. Still working on this.

**Comment by terminalmage** [comment]
The winrepo runner uses the ``force`` argument to the ``git.latest`` state. The problem with this is that this argument is poorly-named. It seems like it should just mimic the various git subcommands' ``--force`` option, but in reality this just forces the `git.latest` state to clone into a directory that doesn't already house a git repository, deleting its contents first.

There is a ``force_reset`` option to ``git.latest`` that would probably be better-suited for this. However, there is a better option to fix this, one which will also stop us from needing to depend on the ``git.latest`` state. I have recently consolidated the gitfs code into a class so that code could be shared by gitfs and git_pillar. The code I have in salt.utils.gitfs would be perfect for winrepo, as it would allow us to fetch and then reset the git checkout (which the recently-rewritten git_pillar is already doing). The amount of code needed to handle winrepo would be pretty small.

I'll look at adding another subclass in salt.utils.gitfs to handle winrepo operations. There are a few caveats to this approach, though:

1. Making this modification will change winrepo from depending on the git CLI to depending on either GitPython or pygit2. I can however check to see if either GitPython or pygit2 is installed, and if not then fall back to the old code, along with a deprecation warning. So this isn't really so much an issue as it is something I'll have to be careful to remember to do.

2. We'd need to remind people upgrading to the release with this fix (via the release notes) to clear out their winrepo cache just to be sure that there are no weird conflicts.

3. This fix could go in no earlier than the 2015.8 release cycle, since that is where the gitfs code was refactored.

**Comment by eliasp** [comment]
@terminalmage Great news, thanks a lot!

I don't see a problem not relying on the git CLI anymore and to use GitPython or pygit2 instead. Calling CLI programs and parsing their output is a bad practice IMHO anyways.

Isn't there a better way to clean out the winrepo cache? Usually, cache data should have some versioning data (e.g. the application release) in their metadata to be able to do these kind of things automagically. I don't know how the winrepo cache data are built internally, but I feel like this would cause quite a backlash in terms of issue reports (even when documenting it in the release notes).

**Comment by terminalmage** [comment]
Well, in this case it isn't a true cache, it's literally just a git checkout.

**Comment by terminalmage** [comment]
@eliasp @UtahDave I've added the gitfs code in my fork, in the ``winrepo`` branch: https://github.com/terminalmage/salt/tree/winrepo

If you're able to test, I'd appreciate it.

Note: This is based off of the 2015.8 branch.

**Comment by terminalmage** [comment]
This was fixed in #26148, closing.
